src/Exercism.Representers.CSharp/Normalization/NormalizeDictionaryInitialization.cs

```diff
@@ -0,0 +1,180 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
+using Serilog;
+
+namespace Exercism.Representers.CSharp.Normalization
+{
+    public class NormalizeDictionaryInitialization : CSharpSyntaxRewriter
+    {
+        private class InvalidKindException : Exception
+        {
+            public SyntaxKind Kind;
+        }
+
+        public override SyntaxNode VisitInitializerExpression(InitializerExpressionSyntax node)
+        {
+            SyntaxNode DefaultVisit() => base.VisitInitializerExpression(node);
+
+            try
+            {
+                if (!IsDictionary(node))
+                {
+                    return DefaultVisit();
+                }
+
+                var initializerExtractionResults
+                    = node.Kind() switch
+                    {
+                        SyntaxKind.CollectionInitializerExpression
+                        => ExtractInitializersWithCollectionSyntax(node),
+                        SyntaxKind.ObjectInitializerExpression
+                        => ExtractInitializersWithObjectSyntax(node),
+                        _
+                        => throw new InvalidKindException {Kind = node.Kind()}
+                    };
+
+                if (!initializerExtractionResults.Success)
+                {
+                    return DefaultVisit();
+                }
+
+                var visitedInitializerSyntaxNodePairs = initializerExtractionResults.InitializerSyntaxNodePairs
+                    .Select(initializer => new KeyValuePair<SyntaxNode, SyntaxNode>(
+                        this.Visit(initializer.Key),
+                        this.Visit(initializer.Value))
+                    );
+
+                var replacementNode
+                    = BuildReplacementSyntaxWithCollectionInitialization(visitedInitializerSyntaxNodePairs);
+
+                if (replacementNode == default)
+                {
+                    var message
+                        = $"{nameof(NormalizeDictionaryInitialization)}: failed to find a {nameof(InitializerExpressionSyntax)} node in generated dictionary fragment";
+                    throw new NullReferenceException(message);
+                }
+
+                return base.VisitInitializerExpression(replacementNode as InitializerExpressionSyntax);
```

Comment:
Maybe make the return type of `BuildReplacementSyntaxWithCollectionInitialization` return an `InitializerExpressionSyntax`? Then you could do:

```suggestion
                return base.VisitInitializerExpression(replacementNode);
```